Cache volume across restarts

bricedp opened the pull request:
This is a continuation of [https://github.com/librespot-org/librespot/pull/218](https://github.com/librespot-org/librespot/pull/218). Now, volume is cached when the user supplies the `--cache` directory. `--initial-volume` overrides cached volume, which in turn overrides the default hard-coded volume.

Currently, the volume is stored in a plaintext file containing only the (percentage) volume. In the future, it could be expanded into a config.json with other parameters as well, since I've largely copied the way that the credentials are stored in credentials.json.

I don't know if the final commit (passing to `Spirc` and `SpircTask`) is the most elegant way to access the cache from `volume_to_mixer`, so I'm open to any improvements to that aspect.

plietar commented on core/src/volume.rs:
```diff
@@ -0,0 +1,39 @@
+        reader.read_to_string(&mut contents).unwrap();
+        let volume = contents.trim().parse::<i32>().unwrap();
+        if volume < 0 || volume > 100 {
+            panic!("Cached volume must be in the range 0-100");
```
Don't panic, just clamp it to (0, 100)

plietar commented on core/src/volume.rs:
```diff
@@ -0,0 +1,39 @@
+use std::path::Path;
+
+pub struct Volume {
+    pub volume: i32,
```
I know it's a bit inconsistent in other places, but use a u16 for volume.

plietar commented on connect/src/spirc.rs:
```diff
@@ -193,7 +196,13 @@ fn calc_logarithmic_volume(volume: u16) -> u16 {
 }
 
-fn volume_to_mixer(volume: u16, linear_volume: bool) -> u16 {
+fn volume_to_mixer(volume: u16, linear_volume: bool, cache: Option<Cache>) -> u16 {
```
This function shouldn't have side effects. Instead save_volume should be called by whoever also calls this method.

You could make a new private `set_volume` on Source, which updates the volume for all of `self.device`, `self.mixer` and `self.cache` and call that everytime.

plietar commented on connect/src/spirc.rs:
```diff
@@ -208,6 +217,7 @@ impl Spirc {
         session: Session,
         player: Player,
         mixer: Box<Mixer>,
+        cache: Option<Cache>,
```
The cache can already be retrieved from the session.

plietar commented on core/src/volume.rs:
```diff
@@ -0,0 +1,39 @@
+use std::io::{Read, Write};
+use std::path::Path;
+
+pub struct Volume {
```
Add `#[derive(Copy, Clone, Debug)]`

plietar commented on core/src/cache.rs:
```diff
@@ -52,6 +53,23 @@ impl Cache {
+        Volume::from_file(path)
+    }
+
+    pub fn save_volume(&self, volume: &Volume) {
```
Once `Volume` implements `Copy`, you can pass it by value here instead of reference.

bricedp commented:
Thanks for the comments @plietar, they were helpful and definitely cleaned up the code. I moved things into a new function within SpircTask that has access to the session. I also fixed a (lack of) rounding error when converting volume from 0..0xFFFF to 0..100 that sometimes caused the cached volume to be decremented by one percent on the next load. Let me know if anything needs editing.

sashahilton00 commented:
@plietar just waiting on your signoff before merging.

plietar commented on src/main.rs:
```diff
@@ -204,28 +204,38 @@ fn setup(args: &[String]) -> Setup {
+            (volume as i32 * 0xFFFF / 100) as u16
         })
-        .unwrap_or(0x8000);
+        .unwrap_or(default_volume);
```
Untested, but this should work and remove the need for the `default_volume` and `cached_volume` temporaries.
```rust
let initial_volume = matches
     .opt_str("initial-volume")
     .map(|volume| {
        let volume = volume.parse::<u16>().unwrap();
        if volume > 100 {
            panic!("Initial volume must be in the range 0-100");
        }
        (volume as i32 * 0xFFFF / 100) as u16
     })
     .or_else(|| cache.as_ref().and_then(Cache::volume));
     .unwrap_or(0x8000);
```

plietar commented on connect/src/spirc.rs:
```diff
@@ -724,6 +730,13 @@ impl SpircTask {
+    fn cache_volume(&self, volume: u16) {
+        if self.session.cache().is_some() {
+            let vol = Volume { volume: volume };
+            self.session.cache().as_ref().unwrap().save_volume(vol);
```
You can avoid the `is_some` and `unwrap` by doing:
```rust
if let Some(cache) = self.session.cache() {
    cache.save_volume(Volume { volume })
}
```

plietar commented on core/src/volume.rs:
```diff
@@ -0,0 +1,40 @@
+    pub volume: u16,
+}
+
+impl Volume {
```
Why not just write the raw 0..0xFFFF value to the file? This isn't meant to be modified by other programs, so it's simpler that way.

bricedp replied:
That was originally to mirror `--initial-volume`, which is represented as a percent. Given that users never need to access the cached value, I can remove that conversion.

plietar commented on connect/src/spirc.rs:
```diff
@@ -724,6 +730,13 @@ impl SpircTask {
         cs.send();
     }
+
+    fn cache_volume(&self, volume: u16) {
```
Change this to `set_volume` and add to this function:
```rust
self.device.set_volume(volume as u32);
self.mixer.set_volume(volume_to_mixer(volume, self.linear_volume));
```

This way the calls can be removed from the other places, since we always need to update all three values

plietar commented on connect/src/spirc.rs:
```diff
@@ -233,11 +234,11 @@ impl Spirc {
+        let volume = config.volume;
         let linear_volume = config.linear_volume;
 
         let device = initial_device_state(config, volume);
```
If you make `SpircTask::set_volume` write to device you can remove the volume argument from this function.
Set the device volume (inside `initial_device_state`) with 0, and overwrite it later when you call `task.set_volume`

bricedp commented:
Thanks @plietar, I've addressed your comments in these latest commits; again let me know if any changes need to be made.

plietar commented:
Looks good.
Thanks!
